### ISSUE TITLE
The loss turned into NaN during training

### ISSUE THREAD

**Issue by eeyrw** [cfzd/Ultra-Fast-Lane-Detection#26]
![图片](https://user-images.githubusercontent.com/2088817/86983578-ca49c280-c1be-11ea-840a-9c9a02c74d4d.png)
The orange line is the result which trained with following cfg:
```

Config (path: configs/culane.py): {'dataset': 'CULane', 'data_root': '/data/home/xxxx/dataset/CULane', 'epoch': 50, 'batch_size': 128, 'optimizer': 'SGD', 'learning_rate': 0.1, 'weight_decay': 0.0001, 'momentum': 0.9, 'scheduler': 'multi', 'steps': [25, 38], 'gamma': 0.1, 'warmup': 'linear', 'warmup_iters': 695, 'use_aux': True, 'griding_num': 200, 'backbone': '18', 'sim_loss_w': 0.0, 'shp_loss_w': 0.0, 'note': 'First train on CULane', 'log_path': './train_log', 'finetune': None, 'resume': None, 'test_model': None, 'test_work_dir': './tmp'}

```
The blue line is the result which trained with following cfg:
```

Config (path: configs/culane.py): {'dataset': 'CULane', 'data_root': '/data/home/xxxx/dataset/CULane', 'epoch': 50, 'batch_size': 128, 'optimizer': 'SGD', 'learning_rate': 0.1, 'weight_decay': 0.0001, 'momentum': 0.9, 'scheduler': 'multi', 'steps': [25, 38], 'gamma': 0.1, 'warmup': 'linear', 'warmup_iters': 695, 'use_aux': True, 'griding_num': 200, 'backbone': '18', 'sim_loss_w': 1, 'shp_loss_w': 1, 'note': 'Train on CULane with other loss', 'log_path': './train_log', 'finetune': None, 'resume': None, 'test_model': None, 'test_work_dir': './tmp'}

```
The difference between cfgs is that the orange one do not applied sim_loss and shp_loss ( the weight of them are set to 0.0 ) and blue one applied with weight 1.0 and 1.0 ( which used in the paper ).  Such kind of difference made loss ill. According my observation, the loss gone into mad in second epoch and here are some log from stdout:
```
.......
 100%|█████████▉| 694/695 [42:47<00:04,  4.02s/it, data_time=1.159, iou=0.194, loss=nan                                                    100%|█████████▉| 694/695 [42:48<00:04,  4.02s/it, data_time=0.055, iou=0.194, loss=nan                                                    100%|██████████| 695/695 [42:48<00:00,  2.96s/it, data_time=0.055, iou=0.194, loss=nan                                                    100%|██████████| 695/695 [42:48<00:00,  3.70s/it, data_time=0.055, iou=0.194, loss=nan                                                    , net_time=0.213, top1=0.001, top2=0.002, top3=0.003]
  0%|          | 0/695 [00:14<?, ?it/s, data_time=13.690, iou=0.194, loss=nan, net_tim                                                      0%|          | 1/695 [00:14<2:51:37, 14.84s/it, data_time=13.690, iou=0.194, loss=na                                                      0%|          | 1/695 [00:16<2:51:37, 14.84s/it, data_time=0.043, iou=0.194, loss=nan                                                      0%|          | 2/695 [00:16<2:04:03, 10.74s/it, data_time=0.043, iou=0.194, loss=nan                                                      0%|          | 2/695 [00:17<2:04:03, 10.74s/it, data_time=0.043, iou=0.195, loss=nan                                                      0%|          | 3/695 [00:17<1:30:50,  7.88s/it, data_time=0.043, iou=0.195, loss=nan                                                      0%|          | 3/695 [00:18<1:30:50,  7.88s/it, data_time=0.059, iou=0.195, loss=nan  
......
```
Is that gradient explosion? But it seems quite good in first epoch.  Furthermore, I use single tesla V100 to train and batchsize is set to 128 according former [issue](https://github.com/cfzd/Ultra-Fast-Lane-Detection/issues/6)

**Comment by cfzd** [comment]
@eeyrw 
I think you can try two options:
- decrease the loss weight of sim_loss and shp_loss 
- use Adam optimizer

The structural loss seems not very stable when using SGD optimizer.

**Comment by eeyrw** [comment]
@cfzd 
Unfortunately, Adam optimizer and decrease weight of sim_loss and shp_loss ( 0,7, 0.5 ) made situation even worse.
![图片](https://user-images.githubusercontent.com/2088817/87102964-32afa700-c286-11ea-9026-2806ed1d424a.png)

Config (path: configs/culane.py): {'dataset': 'CULane', 'data_root': '/data/home/ywen/yrw/dataset/CULane', 'epoch': 50, 'batch_size': 128, 'optimizer': 'Adam', 'learning_rate': 0.1, 'weight_decay': 0.0001, 'momentum': 0.9, 'scheduler': 'multi', 'steps': [25, 38], 'gamma': 0.1, 'warmup': 'linear', 'warmup_iters': 695, 'use_aux': True, 'griding_num': 200, 'backbone': '18', **'sim_loss_w': 0.7, 'shp_loss_w': 0.5,** 'note': 'Train on CULane with loss wght 0.7 0.5 and Adam Optzr', 'log_path': './train_log', 'finetune': None, 'resume': None, 'test_model': None, 'test_work_dir': './tmp'}

**Comment by cfzd** [comment]
@eeyrw 
Maybe you can try grad clip.
```
loss.backward()
torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
optimizer.step()
```
The optimization on the CULane dataset is relatively harder than the one of Tusimple according to reasons in [this issue](https://github.com/cfzd/Ultra-Fast-Lane-Detection/issues/8).

**Comment by eeyrw** [comment]
@cfzd 
I simply give up training on culane with these losses... According that issue, the sim_loss and shp_loss are not used in the network trained on culane. So did you successfully train on culane with these losses?

**Comment by cfzd** [comment]
@eeyrw 
The results reported in our paper are trained with these losses. This is why the results are slightly different.

**Comment by Yansslc** [comment]
Hi, I met the same problem while trainning on culane with default settings in 'config/culane.py'，but by decrease the loss weight of sim_loss and shp_loss both to 0.5 and use Adam optimizer, the train just going well, maybe you can try that again?

**Comment by eeyrw** [comment]
@Yansslc Could you post the detail parameter such as learning rate and batch size ? I want to train it again.

**Comment by Yansslc** [comment]
> @Yansslc Could you post the detail parameter such as learning rate and batch size ? I want to train it again.

I have just finished several trainning with different settings, sorry that I found the loss would not restrained if 'sim_loss_w' and 'shp_loss_w' are not set as zero. Besides, setting both two loss weights as 1.0 as described in the paper doesn't seems reasonable, quite doubt about the results now.

**Comment by eeyrw** [comment]
I tried it again with SGD and sim_loss_w: 0.7 and shp_loss_w: 0.5.
Config (path: configs/culane_with_otherloss.py): {'dataset': 'CULane', 'data_root': '/data/home/ywen/yrw/dataset/CULane', 'epoch': 50, 'batch_size': 128, 'optimizer': 'SGD', 'learning_rate': 0.01, 'weight_decay': 0.0001, 'momentum': 0.9, 'scheduler': 'multi', 'steps': [25, 38], 'gamma': 0.1, 'warmup': 'linear', 'warmup_iters': 695, 'use_aux': True, 'griding_num': 200, 'backbone': '18', 'sim_loss_w': 0.7, 'shp_loss_w': 0.5, 'note': 'Train on CULane with loss wght 0.7 0.5 and SGD Optzr lr/10', 'log_path': './train_log', 'finetune': None, 'resume': None, 'test_model': None, 'test_work_dir': './tmp'}

![图片](https://user-images.githubusercontent.com/2088817/88256361-b9846b00-cced-11ea-9162-1d4a37689992.png)
![图片](https://user-images.githubusercontent.com/2088817/88256397-d28d1c00-cced-11ea-91d0-23b182d26443.png)

The train result is not ideal.

**Comment by Yansslc** [comment]
> I tried it again with SGD and sim_loss_w: 0.7 and shp_loss_w: 0.5.
> Config (path: configs/culane_with_otherloss.py): {'dataset': 'CULane', 'data_root': '/data/home/xxxx/dataset/CULane', 'epoch': 50, 'batch_size': 128, 'optimizer': 'SGD', 'learning_rate': 0.01, 'weight_decay': 0.0001, 'momentum': 0.9, 'scheduler': 'multi', 'steps': [25, 38], 'gamma': 0.1, 'warmup': 'linear', 'warmup_iters': 695, 'use_aux': True, 'griding_num': 200, 'backbone': '18', 'sim_loss_w': 0.7, 'shp_loss_w': 0.5, 'note': 'Train on CULane with loss wght 0.7 0.5 and SGD Optzr lr/10', 'log_path': './train_log', 'finetune': None, 'resume': None, 'test_model': None, 'test_work_dir': './tmp'}
> 
> ![图片](https://user-images.githubusercontent.com/2088817/88256361-b9846b00-cced-11ea-9162-1d4a37689992.png)
> ![图片](https://user-images.githubusercontent.com/2088817/88256397-d28d1c00-cced-11ea-91d0-23b182d26443.png)
> 
> The train result is not ideal.

The relation loss goes even higher?

**Comment by eeyrw** [comment]
@Yansslc I meet such kind of situation frequently... It seems that the network is trapped to the local minimum.

**Comment by Yansslc** [comment]
@eeyrw I tried to reduce the learning rate to 0.001, with  sim_loss_w: 0.5 and shp_loss_w: 0.1, the training is in progress and it seems good for now, I'll update the result after trainning finished if it works.

**Comment by normandra** [comment]
@Yansslc Hi any update on your training result?

**Comment by longzeyilang** [comment]
@normandra Hi any update on your training result?
